### ISSUE TITLE
Batch size set to one during model initialization

### ISSUE THREAD

**Issue by f180n4cc1** [ultralytics/yolov5#8722]
### Search before asking

- [X] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hello everybody.

I've encountered a problem during experimentation with Yolov5. I've have been tampering with adding SKNet like attention mechanism in Yolov5. The issue is that at one point in the attention module I need to use global average pool over the feature space and produce a feature vector that is (batch size, number of channels, 1, 1) that then needs to be batch normalized. When said model is initialized in train.py, an error occurs, since the batch size after model creation is set to 1. I've traced the problem to the __init__ function of Model class in yolo.py, or more precisely to then block of if isinstance(m, Detect). In that block we have a call to self.forward method with dummy tensors from torch.zeros with shape (1, ch, s, s). If I change the batch size to anything other than 1 everything works, which makes sense since batch normalization needs to work on batch size that is greater then 1 during training, in order to calculate mean and variance. My question is then if this is a valid modification, or is there something that I'm missing?

**Comment by glenn-jocher** [comment]
@f180n4cc1 batch normalization works at any batch size including batch size 1, i.e. see example below, so this is likely related to your custom modifications rather than anything in master. If you find any reproducible bugs in master please let us know though.

<img width="1362" alt="Screen Shot 2022-07-26 at 12 58 26 PM" src="https://user-images.githubusercontent.com/26833433/180990492-a7de84bf-463f-47a2-ac01-60a6500f64db.png">

**Comment by f180n4cc1** [comment]
@[glenn-jocher](https://github.com/glenn-jocher). Thanks for the quick reply. Batch normalization does work with batch size 1 on its own, but only if other dimensions are not 1. In my case I need to use global average pooling before batch normalization, that creates a tensor whose shape is (batch_size, num_channels, 1, 1), and if in that instance the batch size is 1, the procedure fails. The easiest way to reproduce it is to change Conv model's forward pass like in this code snippet
`class Conv(nn.Module):
    # Standard convolution
    def __init__(self, c1, c2, k=1, s=1, p=None, g=1, act=True):  # ch_in, ch_out, kernel, stride, padding, groups
        super().__init__()
        self.conv = nn.Conv2d(c1, c2, k, s, autopad(k, p), groups=g, bias=False)
        self.bn = nn.BatchNorm2d(c2)
        self.act = nn.SiLU() if act is True else (act if isinstance(act, nn.Module) else nn.Identity())

    def forward(self, x):
        x = self.conv(x)
        x1 = torch.nn.AdaptiveAvgPool2d((1, 1))(x)
        x1 = self.bn(x1)
        x = torch.mul(x, x1)
        x = self.act(x)
        return self.act(self.bn(self.conv(x)))`
Now if you try to run train with yolov5s.yaml, it will fail in the same way as my custom implementation fails.
edit; code pasting into snippet failed

**Comment by glenn-jocher** [comment]
@f180n4cc1 got it. If your custom architecture/ordering is causing errors then I'd revisit the design of your custom architecture.

**Comment by f180n4cc1** [comment]
@glenn-jocher Thanks, will do.

**Comment by glenn-jocher** [comment]
@f180n4cc1 you're welcome! Good luck with your custom architecture, and feel free to reach out if you have any other questions.
